### PR TITLE
fix(config): reject size env values with trailing units and normalize S3_FORCE_PATH_STYLE

### LINKED ISSUE #1701: [Bug]: size env vars silently truncate values with units (MAX_DEPLOYMENT_FILE_BYTES=100mb becomes 100 bytes) and S3_FORCE_PATH_STYLE only matches lowercase 'false'
### What happened?

Two env-parsing inconsistencies in `backend/src/infra/config/app.config.ts` with real operational consequences. Both verified against current `main`.

**1. Byte-size knobs accept trailing garbage and silently truncate.**

`parseEnvInt` (`app.config.ts:109-115`) uses `parseInt(val, 10)`, so `parseInt("100mb") === 100` sails through the `isNaN/<=0` guard. The byte-valued knobs `MAX_DEPLOYMENT_TOTAL_BYTES` (`:228`) and `MAX_DEPLOYMENT_FILE_BYTES` (`:232`) route through it, and `MAX_FILE_SIZE` (`:176-178`) has its own inline `parseInt` with the same flaw.

The trap is that the *sibling* knob `MAX_JSON_BODY_SIZE=100mb` legitimately takes unit strings (`:174`, Express body-parser format). An operator who writes `MAX_DEPLOYMENT_FILE_BYTES=100mb` by analogy gets a **100-byte** limit: every deployment file is then rejected at `deployment.service.ts:792-794` with `"exceeds the maximum size of 100 bytes"` — an error that echoes the truncated number, not what the operator typed. `MAX_FILE_SIZE=10mb` similarly caps multer at 10 bytes (`middlewares/upload.ts:17,23`), rejecting all uploads.

The strict parser for exactly this class already exists: `parseEnvBytes` (`:125-134`) rejects non-`/^\d+$/` input — but only `S3_MAX_OBJECT_SIZE_BYTES` uses it. So `S3_MAX_OBJECT_SIZE_BYTES=100mb` safely falls back to the default while `MAX_DEPLOYMENT_FILE_BYTES=100mb` bricks deployments.

**2. `S3_FORCE_PATH_STYLE` is an exact-match string comparison.**

`s3ForcePathStyle: process.env.S3_FORCE_PATH_STYLE !== 'false'` (`:215`) — no trim, no lowercase, unlike the repo's own `parseEnvBool` (`:118-121`). Setting `False`, `FALSE`, or ` false ` keeps path-style **on**. The knob's only documented purpose (comment at `s3.provider.ts:105`) is disabling path-style for Tencent COS / Aliyun OSS, which require virtual-hosted-style — so the documented action silently fails for the most common capitalizations, breaking storage for exactly the providers the knob exists to support.

### Proposed fix

- Route all pure-numeric size knobs through the strict `parseEnvBytes`-style validation (or make `parseEnvInt` reject inputs where trailing characters were discarded), with a warning log on fallback.
- `s3ForcePathStyle`: normalize via trim + lowercase (default remains `true` when unset).

### Acceptance criteria

- `MAX_DEPLOYMENT_FILE_BYTES=100mb` → default value + warning, not 100 bytes; same for `MAX_DEPLOYMENT_TOTAL_BYTES`, `MAX_FILE_SIZE`.
- `S3_FORCE_PATH_STYLE=False|FALSE| false ` all disable path-style; unset still defaults to `true`.
- Unit tests cover `"100abc"`, `"100mb"`, `" 100 "`, and the boolean case variants.

Happy to take this.

### PR DESCRIPTION
**Pull request by ayaangazali**
## Summary

Closes #1701.

Two env-parsing bugs in `backend/src/infra/config/app.config.ts`:

1. **Size env vars silently truncated unit suffixes.** `parseInt`-based parsing turned `MAX_DEPLOYMENT_FILE_BYTES=100mb` into `100` — a 100-**byte** limit instead of the intended 100MB, with no warning. Same for the other `*_BYTES`/size vars. Values with trailing units (or any trailing garbage) are now rejected with a clear startup error naming the variable and the accepted format, instead of silently misconfiguring the limit.

2. **`S3_FORCE_PATH_STYLE` only matched lowercase `'false'`.** `FALSE`/`False`/`0` all fell through to path-style enabled. Boolean parsing is now case-insensitive and consistent with the documented values.

## How did you test this change?

- New unit tests covering: unit-suffixed size values rejected (`100mb`, `1g`, whitespace variants), plain numeric values accepted, boundary values, and `S3_FORCE_PATH_STYLE` case-insensitivity (`FALSE`, `False`, `false`, `0`, `true` variants) — `backend/tests/unit/app.config.test.ts` (73 tests in the file, all green).
- Full backend suite: 1811 passed / 17 skipped.
- `npx turbo run typecheck lint --filter=insforge-backend` clean.

---
## Summary by cubic
Rejects unit-suffixed byte-count env vars and normalizes opt-out S3 booleans so only "false" (any case, trimmed) disables them. Prevents accidental 100-byte limits from inputs like "100mb" and keeps path-style/presigned URLs on unless explicitly turned off. Fixes #1701.

- Strict byte limits: `MAX_FILE_SIZE`, `MAX_DEPLOYMENT_TOTAL_BYTES`, `MAX_DEPLOYMENT_FILE_BYTES`, and `S3_MAX_OBJECT_SIZE_BYTES` now accept only trimmed, digit-only positive integers; invalid values warn and fall back. `S3_MAX_OBJECT_SIZE_BYTES` still clamps to 5 GiB and now warns on invalid input.
- Opt-out S3 booleans: `S3_FORCE_PATH_STYLE` and `S3_USE_PRESIGNED_URLS` treat only `false` (any case, trimmed) as false; all other values keep the default `true`.

**Migration**
- Replace unit-suffixed values with plain byte counts for the size env vars above (e.g., use 104857600 instead of "100mb").
- To disable S3 path-style or presigned URLs, set `S3_FORCE_PATH_STYLE=false` or `S3_USE_PRESIGNED_URLS=false`.

<sup>Written for commit 7e0388581b358abfdf1d5eda043abf3aa359734e. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/InsForge/InsForge/pull/1712?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

> [!NOTE]
> ### Reject size env vars with trailing units and normalize `S3_FORCE_PATH_STYLE` parsing
> - Adds `parseStrictEnvInt` in [app.config.ts](https://github.com/InsForge/InsForge/pull/1712/files#diff-24b06be8fbc5ad753cf4ae7c1e6d9c70c7740c17331319f8eb0fc50b559527fb) to validate byte-count env vars as digits-only integers, emitting a `console.warn` and returning a fallback on invalid input (e.g. `100mb` or `1_000`).
> - Adds `isEnvOptOut` to normalize opt-out boolean env vars (`S3_FORCE_PATH_STYLE`, `S3_USE_PRESIGNED_URLS`), treating case/whitespace variants of `'false'` as false and all other values as true.
> - Updates `parseEnvBytes` to delegate to `parseStrictEnvInt`, retaining the AWS single-PUT ceiling clamp.
> - Behavioral Change: `MAX_FILE_SIZE`, `S3_MAX_OBJECT_SIZE_BYTES`, and deployment byte limit env vars previously accepted unit-suffixed values (e.g. `500mb`) silently; they now warn and fall back to defaults.
>
> 
>
> <sup><a href="https://app.macroscope.com">Macroscope</a> summarized 7e03885.</sup>
> 
>

## Summary by CodeRabbit

* **Bug Fixes**
  * Strengthened validation for server file size and deployment/S3 byte-size limits to accept only positive integers/byte counts, rejecting unit-suffixed or trailing-garbage inputs with safe fallbacks.
  * Improved warning behavior: invalid values warn and fall back, while missing/empty values fall back silently.
  * Enhanced normalization for the S3 path-style setting to correctly interpret case/whitespace and disabled inputs (including `0`, `no`, `off`).

* **Tests**
  * Expanded unit coverage for strict parsing rules, warning emission, and clamping to the AWS single-PUT maximum for large S3 upload sizes.

### CODE REVIEW

File: backend/src/infra/config/app.config.ts [hunk 1]
```diff
@@ -115,6 +115,33 @@ function parseEnvInt(val: string | undefined, fallback: number): number {
...
+  val: string | undefined,
+  fallback: T
+): number | T {
+  if (!val) return fallback;
+  const trimmed = val.trim();
+  const parsed = /^\d+$/.test(trimmed) ? Number(trimmed) : NaN;
+  if (!Number.isSafeInteger(parsed) || parsed <= 0) {
+    console.warn(
+      `Invalid ${name} value "${val}": expected a plain positive integer (bytes, no unit suffix); falling back to default.`
+    );
```
[the remark this answers is not in the document]
Comment (ayaangazali) [reply backend/src/infra/config/app.config.ts:152]: Fixed in f32e5113 — the check is now split: non-`/^\d+$/` input keeps the "no unit suffix" hint, while pure-digit non-positive/unsafe values get "must be a positive integer" instead. Regression test added asserting `MAX_FILE_SIZE=0` produces the new message without the unit-suffix wording.

File: backend/src/infra/config/app.config.ts [hunk 2]
```diff
@@ -212,7 +236,7 @@ export function loadConfig(): AppConfig {
       s3EndpointUrl: process.env.S3_ENDPOINT_URL || undefined,
       // Default true (MinIO etc.). Set S3_FORCE_PATH_STYLE=false for providers
       // that require virtual-hosted-style addressing (Tencent COS, Aliyun OSS).
-      s3ForcePathStyle: process.env.S3_FORCE_PATH_STYLE !== 'false',
+      s3ForcePathStyle: (process.env.S3_FORCE_PATH_STYLE ?? '').trim().toLowerCase() !== 'false',
```
[the remark this answers is not in the document]
Comment (ayaangazali) [reply backend/src/infra/config/app.config.ts]: Fixed in f32e5113 — `'0'` (any case/whitespace) now disables path style via `!['false', '0'].includes(...)`, matching what the PR description promised. Tests cover `'0'` and `' 0 '`.

[the remark this answers is not in the document]
Comment (ayaangazali) [reply backend/src/infra/config/app.config.ts]: Fixed in f32e5113 — implementation now uses `!['false', '0'].includes((...).trim().toLowerCase())` and the test file gained a dedicated case covering `'0'` and `' 0 '` (both disable path style).

[the remark this answers is not in the document]
Comment (ayaangazali) [reply backend/src/infra/config/app.config.ts]: Fixed in f32e5113 — `'0'` is now in the false-list alongside `'false'`, with tests for `'0'` / `' 0 '`.

File: backend/src/infra/config/app.config.ts [hunk 3]
```diff
@@ -115,6 +115,39 @@ function parseEnvInt(val: string | undefined, fallback: number): number {
...
+ * Strict positive-integer parser for pure byte-count knobs (MAX_FILE_SIZE,
+ * MAX_DEPLOYMENT_*_BYTES). parseInt() silently truncates trailing garbage
+ * ("100mb" → 100), which is a trap here because the sibling body-size knobs
+ * (MAX_JSON_BODY_SIZE) legitimately accept unit suffixes. Mirrors the
+ * /^\d+$/ strictness of parseEnvBytes and warns on rejection so operators
+ * can see why their value was ignored. Uses console.warn because the shared
+ * logger (utils/logger.ts) imports appConfig, so it cannot be used at
+ * config-load time.
+ */
+function parseStrictEnvInt<T extends number | undefined>(
```
Comment (Fermionic-Lyu) [inline backend/src/infra/config/app.config.ts:142]: This is a near-duplicate of `parseEnvBytes` below — same `/^\d+$/` and safe-integer checks, minus the trim/warn and the 5 GiB cap. Build `parseEnvBytes` on top of this so `S3_MAX_OBJECT_SIZE_BYTES` behaves the same way: today `S3_MAX_OBJECT_SIZE_BYTES=100mb` still falls back silently, and it rejects whitespace-padded values this parser accepts.

File: backend/src/infra/config/app.config.ts [hunk 4]
```diff
@@ -210,9 +240,12 @@ export function loadConfig(): AppConfig {
       awsAccessKeyId: process.env.AWS_ACCESS_KEY_ID || undefined,
       awsSecretAccessKey: process.env.AWS_SECRET_ACCESS_KEY || undefined,
       s3EndpointUrl: process.env.S3_ENDPOINT_URL || undefined,
-      // Default true (MinIO etc.). Set S3_FORCE_PATH_STYLE=false for providers
-      // that require virtual-hosted-style addressing (Tencent COS, Aliyun OSS).
-      s3ForcePathStyle: process.env.S3_FORCE_PATH_STYLE !== 'false',
+      // Default true (MinIO etc.). Set S3_FORCE_PATH_STYLE=false (or 0) for
+      // providers that require virtual-hosted-style addressing (Tencent COS,
+      // Aliyun OSS).
+      s3ForcePathStyle: !['false', '0'].includes(
```
Comment (Fermionic-Lyu) [inline backend/src/infra/config/app.config.ts]: This falsy-list is asymmetric with `parseEnvBool`'s truthy-list (`1`/`true`/`yes`/`on`): `S3_FORCE_PATH_STYLE=no` or `off` silently keeps path style on — the same trap this PR fixes for `0`. Accept `no`/`off` here too (or add a default-true counterpart to `parseEnvBool`) so the file's two boolean conventions agree.

**Review by Fermionic-Lyu — CHANGES_REQUESTED** [review]
Two consistency fixes, inline.

File: backend/src/infra/config/app.config.ts (hunk 3 above)
Comment (ayaangazali) [reply backend/src/infra/config/app.config.ts:142]: Fixed in a88cfbd1 — `parseEnvBytes` is now just `Math.min(parseStrictEnvInt(name, val, fallback), AWS_MAX_SINGLE_PUT_BYTES)`, so `S3_MAX_OBJECT_SIZE_BYTES` gets the trim + warn behavior (`100mb` now warns instead of silently falling back, whitespace-padded values parse). Tests cover the warn, the trim, and the 5 GiB clamp.

File: backend/src/infra/config/app.config.ts (hunk 4 above)
Comment (ayaangazali) [reply backend/src/infra/config/app.config.ts]: Fixed in a88cfbd1 — falsy list is now `['false', '0', 'no', 'off']` (case/whitespace-insensitive), mirroring `parseEnvBool`'s truthy list. Tests cover `no`/`NO`/`off`/` Off `.

**Comment by ayaangazali** [comment]
@Fermionic-Lyu both consistency fixes are in (a88cfbd1): `parseEnvBytes` now builds on `parseStrictEnvInt` (S3_MAX_OBJECT_SIZE_BYTES gets trim + warn, `100mb` no longer falls back silently) and the falsy list accepts `no`/`off`, mirroring `parseEnvBool`. 79 config tests green, CI green. Threads resolved — ready for re-review whenever you have a minute (fork PRs can't re-request via the API).

**Comment by ayaangazali** [comment]
Rebased onto current main (5fa84a694) and green: typecheck clean, `app.config` tests 82/82.

Status in one line: @Fermionic-Lyu's two consistency asks from Jul 17 (dedupe `parseEnvBytes` onto `parseStrictEnvInt`, and add `no`/`off` to the falsy list) both landed in a88cfbd1.

One thing that may be holding this up mechanically: the PR still shows CHANGES_REQUESTED from that Jul 17 review. As the author of a fork PR I cannot clear that flag or re-request review through the UI, so it stays red until a reviewer takes another look, even though the changes are in. Not urgent, just flagging in case it looked like an open request.

**Comment by ayaangazali** [comment]
Status note so this is not confusing to a reviewer: issue #1701 is currently unassigned. The assignment bot has released me from it six times now, most recently today, and I used the freed slot to claim #1835 for #1839 instead, since that PR was waiting on exactly that.

Nothing about this PR changed. It is rebased onto current main, green, and the two consistency fixes @Fermionic-Lyu asked for on Jul 17 are in (a88cfbd1). I will reclaim #1701 as soon as a slot frees, or happily leave it unassigned if that is simpler.

**Comment by Fermionic-Lyu** [comment]
> Status note so this is not confusing to a reviewer: issue #1701 is currently unassigned. The assignment bot has released me from it six times now, most recently today, and I used the freed slot to claim #1835 for #1839 instead, since that PR was waiting on exactly that.
> 
> Nothing about this PR changed. It is rebased onto current main, green, and the two consistency fixes @Fermionic-Lyu asked for on Jul 17 are in ([a88cfbd](https://github.com/InsForge/InsForge/commit/a88cfbd14be5b1917a885598c786ae17c4d58ee2)). I will reclaim #1701 as soon as a slot frees, or happily leave it unassigned if that is simpler.

Hi, please click that re-request review at your convenience. If you don't do that, the PR never enters our review worflow. Thanks

**Comment by ayaangazali** [comment]
@Fermionic-Lyu I tried, and I cannot. That button is not available to me as an outside contributor, and it is not a UI-only limitation. Both API paths refuse it:

REST `POST /repos/InsForge/InsForge/pulls/1712/requested_reviewers`:
```
{"message": "Not Found", "status": "404"}
```

GraphQL `requestReviews`:
```
{"type": "FORBIDDEN", "message": "ayaangazali does not have the correct permissions to execute `RequestReviews`"}
```

GitHub only allows requesting reviewers with write access to the repo, and a fork-based contributor never has it. So if a PR entering your review workflow depends on the author clicking re-request, an outside PR cannot enter it on its own, which I suspect is not what you intended.

Two things that would work, whichever you prefer:

1. You re-request it, since a reviewer with write access can. I will reply the same day to anything you find.
2. Tell me a signal I *can* send, for example a comment with a set phrase, or re-opening the PR, and I will use that from now on and mention it in CONTRIBUTING so the next contributor is not stuck the same way.

State of this PR meanwhile: rebased on current main, all checks green, and your two consistency asks from Jul 17 are in a88cfbd.

**Comment by ayaangazali** [comment]
Pushed a7e73702e, which corrects two things in my own diff. Both are narrowing, not additions.

**1. I had gone past what the issue asked for.** My change treated `0`, `no` and `off` as falsy for `S3_FORCE_PATH_STYLE`, mirroring `parseEnvBool`'s truthy list. That was the wrong direction for this kind of knob, and the repo already says so: the existing test for `S3_USE_PRESIGNED_URLS` asserts `'0'` keeps the knob **on**, with the comment "Garbage values keep the safe default, matching the S3_FORCE_PATH_STYLE parsing convention". These knobs default ON, so an unrecognised value should keep the default rather than guess. Landing my version would have made that comment false. The actual defect in #1701 was only that the comparison was exact, so `False`, `FALSE` and ` false ` all read as "not false". Normalization is now trim plus lowercase against `'false'` alone, which is exactly the acceptance criteria in #1701.

**2. I had fixed one of the two knobs.** `s3UsePresignedUrls` sits three lines below in `loadConfig` with the same `!== 'false'` comparison, and I left it. It matters more than the one I did fix: turning presigned URLs off is what an operator does when the S3 endpoint is not browser-reachable (bundled MinIO/RustFS on the Docker network, or R2 without POST-policy support), so `S3_USE_PRESIGNED_URLS=False` silently keeps handing clients URLs they cannot fetch. Both knobs now go through one `isEnvOptOut` helper.

Net effect on the diff: the parsing is smaller and the two siblings agree.

```
S3_FORCE_PATH_STYLE=False   -> path style off   (was: on)
S3_USE_PRESIGNED_URLS=False -> proxy mode on    (was: presigned, unchanged from before this PR)
S3_FORCE_PATH_STYLE=0       -> path style on    (safe default, as the convention says)
```

Gates: `app.config` tests 99/99, the two S3 provider suites 35/35, typecheck clean, prettier clean.

@Fermionic-Lyu no action needed on this beyond the re-request question in my previous comment.

**Comment by Fermionic-Lyu** [comment]
> @Fermionic-Lyu I tried, and I cannot. That button is not available to me as an outside contributor, and it is not a UI-only limitation. Both API paths refuse it:
> 
> REST `POST /repos/InsForge/InsForge/pulls/1712/requested_reviewers`:
> 
> ```
> {"message": "Not Found", "status": "404"}
> ```
> 
> GraphQL `requestReviews`:
> 
> ```
> {"type": "FORBIDDEN", "message": "ayaangazali does not have the correct permissions to execute `RequestReviews`"}
> ```
> 
> GitHub only allows requesting reviewers with write access to the repo, and a fork-based contributor never has it. So if a PR entering your review workflow depends on the author clicking re-request, an outside PR cannot enter it on its own, which I suspect is not what you intended.
> 
> Two things that would work, whichever you prefer:
> 
> 1. You re-request it, since a reviewer with write access can. I will reply the same day to anything you find.
> 2. Tell me a signal I _can_ send, for example a comment with a set phrase, or re-opening the PR, and I will use that from now on and mention it in CONTRIBUTING so the next contributor is not stuck the same way.
> 
> State of this PR meanwhile: rebased on current main, all checks green, and your two consistency asks from Jul 17 are in [a88cfbd](https://github.com/InsForge/InsForge/commit/a88cfbd14be5b1917a885598c786ae17c4d58ee2).

It is a UI button at the top right 'Reviewers' section on this page.

**Comment by ayaangazali** [comment]
You will see that button, because you have write access to this repo. I do not, and that is the whole of the difference. There is no Reviewers control in my view of this page.

I re-checked before replying rather than just repeating myself, in case my first attempt was malformed. It was not. With a properly formed JSON body and a token carrying full `repo` scope:

```
$ gh api -X POST repos/InsForge/InsForge/pulls/1712/requested_reviewers \
    --input <(echo '{"reviewers":["Fermionic-Lyu"]}')
{"message":"Not Found","status":"404"}

$ X-OAuth-Scopes: gist, read:org, repo, workflow
```

GitHub returns 404 rather than 403 here because it hides the endpoint's existence from accounts without push access. The GraphQL mutation the button calls says it more directly: `FORBIDDEN: ayaangazali does not have the correct permissions to execute RequestReviews`.

I am not trying to win a point about permissions, and I would rather not spend more of your time on it. What I actually need is small: this PR has been ready since Jul 17, when I addressed both of your consistency asks in a88cfbd. It is rebased on current main and green.

If you click re-request, it takes you a second and I will respond same-day to whatever you find. If it is easier to just look at the diff, it is 74 lines in one file plus tests, and the summary is:

- `MAX_DEPLOYMENT_FILE_BYTES=100mb` currently parses to **100 bytes** and bricks deployments, because the byte knobs route through `parseInt`. They now use a strict parser that rejects unit suffixes and warns.
- `S3_FORCE_PATH_STYLE=False` currently does nothing, because the comparison was exact-match against lowercase `'false'`. Now trimmed and lowercased.
- Same fix applied to `S3_USE_PRESIGNED_URLS`, which had the identical bug three lines below and which I had missed on the first pass.

Either way works for me. I just do not have a way to move it forward from my side.

**Comment by ayaangazali** [comment]
Housekeeping only, no new argument on top of my last comment: brought this up to current main (it was 100 commits behind, merged clean). Re-ran the gates after the merge rather than assuming they still held — `app.config` tests 99/99, backend typecheck clean.

Re-verified the defect is still live on `main` @ 578cda898 instead of trusting a two-week-old check:

```
app.config.ts:288  s3ForcePathStyle:   process.env.S3_FORCE_PATH_STYLE !== 'false'
app.config.ts:294  s3UsePresignedUrls: process.env.S3_USE_PRESIGNED_URLS !== 'false'
app.config.ts:311  maxDeploymentFileBytes: parseEnvInt(process.env.MAX_DEPLOYMENT_FILE_BYTES, ...)
```

So `MAX_DEPLOYMENT_FILE_BYTES=100mb` still parses to 100 bytes, and `S3_FORCE_PATH_STYLE=False` still does nothing.

@Fermionic-Lyu no action needed from you beyond whatever you decide on my previous comment. This is just keeping the branch current so it is not stale whenever you get to it.

**Comment by ayaangazali** [comment]
Process note, since this is the one thing that could still block this PR on a technicality: **#1701 is now assigned to me**, so this PR is linked to a claimed issue like the others. It had been sitting unbacked because the assignment bot released #1701 repeatedly and I had no free slot to reclaim it.

Nothing else changed. Still rebased on current main, 0 behind, no failing checks, and the defect is still live on `main` @ 578cda898:

```
app.config.ts:288  s3ForcePathStyle:   process.env.S3_FORCE_PATH_STYLE !== 'false'
app.config.ts:294  s3UsePresignedUrls: process.env.S3_USE_PRESIGNED_URLS !== 'false'
app.config.ts:311  maxDeploymentFileBytes: parseEnvInt(process.env.MAX_DEPLOYMENT_FILE_BYTES, ...)
```

@Fermionic-Lyu no reply needed on this, it is bookkeeping so the PR is not held up for a reason unrelated to the code.

**Comment by ayaangazali** [comment]
Correcting three of my own review replies on this PR, because they now say the opposite of what the code does.

On Jul 15 I answered the greptile P2, CodeRabbit and cubic P2 threads on `app.config.ts:239` with "Fixed — `'0'` (any case/whitespace) now disables path style". That was true when I wrote it. On Aug 9 I reversed it, and I explained the reasoning in a PR comment but never went back to the inline threads, so they have been contradicting the implementation for over a week. Anyone reading them would reasonably conclude the code is broken.

**Current behaviour, verified on this branch just now:**

```ts
function isEnvOptOut(val: string | undefined): boolean {
  return (val ?? '').trim().toLowerCase() === 'false';
}
...
s3ForcePathStyle:   !isEnvOptOut(process.env.S3_FORCE_PATH_STYLE),
s3UsePresignedUrls: !isEnvOptOut(process.env.S3_USE_PRESIGNED_URLS),
```

So `'0'`, `'no'` and `'off'` do **not** disable these knobs. Only `false` does, case-insensitively and whitespace-tolerantly.

**Why I reversed it**, in short: these knobs default ON, and the repo already had a stated convention that an unrecognised value keeps the safe default rather than guessing. `app.config.test.ts` asserted exactly that for `S3_USE_PRESIGNED_URLS`, with a comment saying "garbage values keep the safe default, matching the S3_FORCE_PATH_STYLE parsing convention". Accepting `0/no/off` would have made that comment false, and it also went past what issue #1701 asked for, which was trim plus lowercase. The real defect was only that the comparison was exact, so `False` and `FALSE` silently did nothing.

If you would rather have the wider falsy list after all, say so and I will add it along with updating that test comment. I am not attached to the narrower version, only to the two not disagreeing.

Nothing else changed. Branch merged up to current main, `app.config` tests 99/99.

**Comment by ayaangazali** [comment]
@Fermionic-Lyu I am going to close and immediately reopen this, because you demonstrated the mechanism I was missing.

On #1952 today you wrote: "Closing to rebase onto `main` [...] and to reopen for a clean read". That is something a fork contributor *can* do, unlike the re-request button, which GitHub only offers to accounts with write access. So it is the signal I have been looking for since Aug 12, and I would not have thought of it if you had not just used it.

Nothing about the diff changes. State before the reopen:

- rebased onto current main, 0 commits behind
- no failing checks
- no unresolved review threads
- `app.config` tests 99/99, backend typecheck clean
- #1701 is assigned to me, so the issue-first request is satisfied

The defect is still live on `main` @ 3ccd05150:

```
app.config.ts:288  s3ForcePathStyle:   process.env.S3_FORCE_PATH_STYLE !== 'false'
app.config.ts:294  s3UsePresignedUrls: process.env.S3_USE_PRESIGNED_URLS !== 'false'
app.config.ts:311  maxDeploymentFileBytes: parseEnvInt(process.env.MAX_DEPLOYMENT_FILE_BYTES, ...)
```

So `MAX_DEPLOYMENT_FILE_BYTES=100mb` still parses to 100 bytes and bricks deployments, and `S3_FORCE_PATH_STYLE=False` still silently does nothing.

If close-and-reopen is not the signal you meant, tell me what is and I will use that instead. I am only trying to get it into your queue, not to churn the PR.

**Comment by ayaangazali** [comment]
The red `Dashboard E2E Tests` here is not from this diff. Pushed an empty commit to request a fresh run, since a contributor cannot re-run a job without write access.

Evidence, from the job API rather than the summary:

```
conclusion = cancelled
started    03:08:28Z
completed  09:08:43Z        <- exactly 6h, the Actions maximum job time

step 5  Install Playwright Chromium : cancelled  03:09:02Z -> 09:08:41Z
step 6  Run dashboard E2E tests     : skipped
```

The browser download hung for 5h59m and the job was killed at the ceiling. The tests never executed, so nothing in the diff could have influenced the result.

It is also transient rather than repo-wide: the same workflow succeeded on `feat/ui-popover-calendar-clock` at 05:43Z, and on this very branch at 22:11Z the previous evening. Only the runs inside a roughly two-hour window were affected.

File: backend/src/infra/config/app.config.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (ayaangazali) [reply backend/src/infra/config/app.config.ts]: Correction to my own reply above, which is now wrong.

I said `'0'` disables path style. That was true when I wrote it on Jul 15, but I reversed it on Aug 9: `isEnvOptOut` now matches an explicit `false` only, case-insensitively and whitespace-tolerantly, so `'0'`, `'no'` and `'off'` keep these knobs ON.

Why: both knobs default ON, and the suite already stated the convention that an unrecognised value keeps the safe default (the `S3_USE_PRESIGNED_URLS` test comment said so, citing the `S3_FORCE_PATH_STYLE` convention). Accepting `0/no/off` would have made that comment false, and it went past what issue #1701 asked for, which was trim plus lowercase. The actual defect was only that the comparison was exact, so `False` and `FALSE` silently did nothing.

Leaving this note in the thread rather than only at PR level, because a resolved thread that says the opposite of the code is worse than no reply. Happy to widen the falsy list if the maintainers prefer it, along with updating that test comment.

[the remark this answers is not in the document]
Comment (ayaangazali) [reply backend/src/infra/config/app.config.ts]: Correction to my own reply above, which is now wrong.

I said `'0'` disables path style. That was true when I wrote it on Jul 15, but I reversed it on Aug 9: `isEnvOptOut` now matches an explicit `false` only, case-insensitively and whitespace-tolerantly, so `'0'`, `'no'` and `'off'` keep these knobs ON.

Why: both knobs default ON, and the suite already stated the convention that an unrecognised value keeps the safe default (the `S3_USE_PRESIGNED_URLS` test comment said so, citing the `S3_FORCE_PATH_STYLE` convention). Accepting `0/no/off` would have made that comment false, and it went past what issue #1701 asked for, which was trim plus lowercase. The actual defect was only that the comparison was exact, so `False` and `FALSE` silently did nothing.

Leaving this note in the thread rather than only at PR level, because a resolved thread that says the opposite of the code is worse than no reply. Happy to widen the falsy list if the maintainers prefer it, along with updating that test comment.

[the remark this answers is not in the document]
Comment (ayaangazali) [reply backend/src/infra/config/app.config.ts]: Correction to my own reply above, which is now wrong.

I said `'0'` disables path style. That was true when I wrote it on Jul 15, but I reversed it on Aug 9: `isEnvOptOut` now matches an explicit `false` only, case-insensitively and whitespace-tolerantly, so `'0'`, `'no'` and `'off'` keep these knobs ON.

Why: both knobs default ON, and the suite already stated the convention that an unrecognised value keeps the safe default (the `S3_USE_PRESIGNED_URLS` test comment said so, citing the `S3_FORCE_PATH_STYLE` convention). Accepting `0/no/off` would have made that comment false, and it went past what issue #1701 asked for, which was trim plus lowercase. The actual defect was only that the comparison was exact, so `False` and `FALSE` silently did nothing.

Leaving this note in the thread rather than only at PR level, because a resolved thread that says the opposite of the code is worse than no reply. Happy to widen the falsy list if the maintainers prefer it, along with updating that test comment.

**Comment by ayaangazali** [comment]
@Fermionic-Lyu status against current `main` (`ccb1703`), with a fresh local verification rather than a restatement.

**Your two consistency fixes from Jul 17 are in** (`a88cfbd`) and both inline threads are resolved. Nothing on this PR is unanswered.

**The defect still reproduces on today's main.** Both halves are still literally present:

```
app.config.ts:125   const parsed = parseInt(val, 10);                        # "100mb" -> 100 bytes
app.config.ts:238   const parsed = parseInt(process.env.MAX_FILE_SIZE || '', 10);
app.config.ts:288   s3ForcePathStyle: process.env.S3_FORCE_PATH_STYLE !== 'false',
app.config.ts:294   s3UsePresignedUrls: process.env.S3_USE_PRESIGNED_URLS !== 'false',
```

**Verified the fix against current main, not against this PR's stale base.** I applied only this PR's two files onto `ccb1703` in a scratch worktree and ran the one affected test file:

```
with the fix:      Test Files 1 passed (1)      Tests 99 passed (99)
source reverted:   Test Files 1 failed (1)      Tests 9 failed | 90 passed (99)
```

The 9 that fail without the fix are the ones that matter, including `rejects MAX_FILE_SIZE with a unit suffix ("100mb") ... instead of truncating to 100 bytes`, `disables path style for case and whitespace variants of false`, and `enters proxy mode for case and whitespace variants of false`. So the tests do fail on unfixed code, which I checked by actually reverting the source rather than assuming.

**On CI and staleness:** the green here is from Aug 18, against a base 75 commits old, so I did not want to offer it as proof. `app.config.ts` and its test were not touched by any of those 75 commits, and the branch is `MERGEABLE`, so a rebase would reset CI without changing anything real. The local run above is the current evidence instead.

**On the re-request review button:** still not available to me. Both API paths refuse it for a fork contributor (`404` on REST `requested_reviewers`, `FORBIDDEN` on GraphQL `requestReviews`), which is why I used close-and-reopen on Aug 18 after you demonstrated it on your own #1952. Nothing has landed since then, so I have not repeated it; that would just be churn.

Nothing outstanding from me. Waiting on review.
